src/lib/support/CHIPSparseArray.h

```diff
@@ -0,0 +1,193 @@
+     *
+     * @param[in] serialized Serialized buffer
+     * @param[in] buflen Length of buffer
+     * @return CHIP_NO_ERROR in case of success, or the error code
+     */
+    CHIP_ERROR Deserialize(const char * serialized, uint16_t buflen)
+    {
+        CHIP_ERROR err     = CHIP_NO_ERROR;
+        uint16_t decodelen = 0;
+        VerifyOrExit(buflen <= MaxSerializedSize(), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+        decodelen = Base64Decode(serialized, buflen, reinterpret_cast<uint8_t *>(mArray));
+        VerifyOrExit(decodelen <= sizeof(uint64_t) * Capacity, err = CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrExit(decodelen % sizeof(uint64_t) == 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+        mNextAvailable = decodelen / static_cast<uint16_t>(sizeof(uint64_t));
+    exit:
+        return err;
+    }
+
+    /**
+     * @brief
+     *   Get the length of string if the array is serialized.
+     */
+    uint16_t SerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * mNextAvailable)); }
+
+    /**
+     * @brief
+     *   Get the maximum length of string if the array was full and serialized.
+     */
+    uint16_t MaxSerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * Capacity)); }
+
+    /**
+     * @brief
+     *   Find a value in the array.
+     *
+     * @param[in] value Value to find
+     * @return index of the value if found, or max length (Capacity) of the array
+     */
+    uint16_t Find(uint64_t value)
+    {
+        uint64_t valueLE = ToLE64(value);
+        for (uint16_t i = 0; i < mNextAvailable; i++)
+        {
+            if (mArray[i] == valueLE)
+            {
+                return i;
+            }
+        }
+
+        return Capacity;
+    }
+
+    /**
+     * @brief
+     *   Insert the value in the array. If the value is duplicate, it
+     *   won't be inserted.
+     *
+     * @return CHIP_NO_ERROR in case of success, or the error code
+     */
+    CHIP_ERROR Insert(uint64_t value)
+    {
+        CHIP_ERROR err = CHIP_ERROR_INTERNAL;
+        uint16_t index = FirstAvailableForUniqueId(value);
+        if (index < Capacity)
+        {
+            mArray[index] = ToLE64(value);
+            if (index == mNextAvailable)
+            {
+                mNextAvailable = static_cast<uint16_t>(index + 1);
+            }
+            err = CHIP_NO_ERROR;
+        }
+        return err;
+    }
+
+    /**
+     * @brief
+     *   Delete the value from the array.
+     */
+    void Remove(uint64_t value)
+    {
+        uint16_t index = Find(value);
+        if (index < Capacity)
+        {
+            mArray[index] = 0;
+            if ((index + 1) == mNextAvailable)
+            {
+                mNextAvailable = index;
+                while (mNextAvailable > 0 && mArray[mNextAvailable - 1] == 0)
+                    mNextAvailable--;
+            }
+        }
+    }
+
+private:
+    uint64_t mArray[Capacity];
```

Comment:
Worth documenting the endianness weirdness around this storage and why we do it that way.